bentoml/yatai/yatai_service_impl.py

```diff
@@ -578,6 +584,79 @@ def ContainerizeBento(self, request, context=None):
                     return ContainerizeBentoResponse(status=Status.INTERNAL(e))
 
-            # pylint: enable=unused-argument
+        def UploadBento(self, request_iterator, context=None):
+            try:
+                with TempDirectory() as temp_dir:
+                    temp_tar_path = os.path.join(
+                        temp_dir, f'{uuid.uuid4().hex[:12]}.tar'
+                    )
+                    with open(temp_tar_path, 'wb+') as file:
+                        for request in request_iterator:
+                            if not request.bento_bundle:
+                                bento_name = request.bento_name
+                                bento_version = request.bento_version
+                            else:
+                                if (
+                                    bento_name == request.bento_name
+                                    and bento_version == request.bento_version
+                                ):
+                                    file.write(request.bento_bundle)
+                                else:
+                                    raise BentoMLException(
+                                        f"Incoming stream request doesn't match "
+                                        f"with initial request info "
+                                        f"{bento_name}:{bento_version} - "
+                                        f"{request.bento_name}:"
+                                        f"{request.bento_version}"
+                                    )
+                        bento_id = f"{bento_name}_{bento_version}"
+                        with lock(self.db, [(bento_id, LockType.WRITE)]) as (sess, _):
+                            bento_pb = self.db.metadata_store.get(
+                                sess, bento_name, bento_version
+                            )
+                            file.seek(0)
+                            if bento_pb:
+                                with tarfile.open(fileobj=file, mode='r') as tar:
+                                    tar.extractall(path=bento_pb.uri.uri)
+                                result_status = Status.OK()
+                            else:
+                                result_status = Status.NOT_FOUND(
+                                    "BentoService `{}:{}` is not found".format(
+                                        bento_name, bento_version
+                                    )
+                                )
+                    return UploadBentoResponse(status=result_status)
+            except BentoMLException as e:
+                logger.error("RPC ERROR UploadBento: %s", e)
+                return UploadBentoResponse(status=e.status_proto)
+            except Exception as e:  # pylint: disable=broad-except
+                logger.error("RPC ERROR UploadBento: %s", e)
+                return UploadBentoResponse(status=Status.INTERNAL())
+
+        def DownloadBento(self, request, context=None):
```

Comment:
Should we add a check here to verify that `self.repository` is a FileSystemRepository? Looks like this RPC method only supports FileSystemRepository right now.

Reply:
Same for the UploadBento rpc above?

Reply:
We checked bento bundle's URI type to be file system before making calling the internal methods for uploading and downloading.

Reply:
That's on the client side, would be better yo verify it on server side as well